---
title: "PCAGenomicSignatures - Contents"
author: "Sehyun Oh"
date: "`r format(Sys.time(), '%B %d, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Information in PCAGenomicSignatures}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    number_sections: no
    toc: yes
    toc_depth: 4
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE, comment = "#>" 
)
```

# Setup

## Install and load package
```{r eval = FALSE}
if (!require("devtools")) {
  install.packages("devtools")
}
devtools::install_github("shbrief/PCAGenomicSignatures")
```

```{r results="hide", message=FALSE, warning=FALSE}
library(PCAGenomicSignatures)
```


## Download PCAmodel
Currently, you can download PCAGenomicSignatures from Google Cloud bucket using 
`PCAGenomicSignatures::getModel` function. This model is built from top 20 PCs of 
536 studies (containing 44,890 samples) containing 13,934 common genes from each 
of 536 study's top 90% varying genes based on thier study-level standard deviation.
There are two versions of this using different gene sets for GSEA-based annotation;
MSigDB C2 (`C2`) and three priors from PLIER package (`PLIERpriors`). In this 
vignette, we are using the `C2` annotated model.

```{r load_model}
if (!file.exists("PCAmodel_C2.rds")) {getModel("C2")} 
PCAmodel <- readRDS("PCAmodel_C2.rds")
PCAmodel
```



## Quick Summary of main information
* model(PCAmodel) : avgLoading matrix   
* studies(PCAmodel) : a list of character vectors, containing studies in each loading   
* gsea(PCAmodel) : a list of gseaResult objects of each loading   
* mesh(PCAmodel) : a list of data frames containing MeSH information associated with each study   
* PCAsummary(PCAmodel) : a list of matrix (3 by 20) containing PCA summary of each study   


# Accessors
## PCclusters/ avgLoadings/ model
Average loadings (= avgLoading = PCclusters) is the main part of GenomicSignatures, 
serving as an index connecting new datasets and the existing database. You can access 
it through `PCAGenomicSignatures::model` (equivalent of `SummarizedExperiment::assay`). 
Rows are genes and columns are avgLoading vectors.

Here, PCAmodel consists of 13,934 genes and 4,764 PCclusters.

```{r}
class(model(PCAmodel))
dim(model(PCAmodel))
model(PCAmodel)[1:4, 1:4]
```

## Metadata
Metadata slot of the PCAGenomicSignatures object contains information related to 
the model buildling.

```{r}
names(metadata(PCAmodel))
```

`cluster` contains the cluster membership of each PCs from training dataset and 
`size` is the integer vector with the length of clusters, containing the size of 
each cluster. `k` is the number of clusters and `n` is the number of top PCs kept
from each study in training dataset. `MeSH_freq` is the information on the frequency
of MeSH terms assigned to training dataset. You can see that MeSH term 'Humans' 
and 'RNA-seq' are top ranked, which is very expected because the training dataset
of this model is Human RNA sequencing. `updateNote` tells you a brief note on the
given model's specification.

```{r}
head(metadata(PCAmodel)$cluster)
head(metadata(PCAmodel)$size)
metadata(PCAmodel)$k
metadata(PCAmodel)$n
head(metadata(PCAmodel)$MeSH_freq)
updateNote(PCAmodel)  
```


## Studies in each loading
You can find which studies are in each cluster using `studies` method. Output is 
a list with the length of clusters, where each element is a character vector containing
the name of studies in each cluster.

```{r}
length(studies(PCAmodel))
studies(PCAmodel)[1:3]
```

You can check which PC from different studies are in PCclusters using `PCinCluster`.

```{r}
PCinCluster(PCAmodel, 2)
```

## Silhouette Width
Silhouette width ranges from -1 to 1 for each observation in your data and can be
interpreted as follows:   
  - Values close to 1 suggest that the observation is well matched to the assigned cluster   
  - Values close to 0 suggest that the observation is borderline matched between two clusters   
  - Values close to -1 suggest that the observations may be assigned to the wrong cluster   
  
Can I use silhouette width to assign the level of validity for avgLoadings?

```{r}
x <- silhouetteWidth(PCAmodel)
summary(x)
```

## GSEA on each loading
```{r}
class(gsea(PCAmodel))
class(gsea(PCAmodel)[[1]])
length(gsea(PCAmodel))
gsea(PCAmodel)[1]
```


## MeSH terms for each study
You can find MeSH terms associated with each study using `mesh(x)` method. Output is 
a list with the length of studies in training dataset, where each element is a 
data frame containing the assigned MeSH terms and the detail of them. The last 
column `bagOfWords` is the frequency of that MeSH term in the whole training dataset.

```{r}
length(mesh(PCAmodel))
mesh(PCAmodel)[1]
```


## PCA summary
You can find the PCA summary of each study using `PCAsummary` method. Output is 
a list with the length of studies, where each element is a matrix containing PCA
summary results with SD/Variance/Cumulative.

```{r}
length(PCAsummary(PCAmodel))
PCAsummary(PCAmodel)[1]
```
